text/0000-native-class-roadmap.md

```diff
@@ -0,0 +1,265 @@
+and spread knowledge of best practices when using native classes.
+
+## Drawbacks
+
+This approach limits what we can do significantly in framework code (no
+inheritance, can't provide defaults, must provide all behavior external to user
+code). There are many established software patterns other than inheritance for
+providing most behavior, so in practice this shouldn't be an issue.
+
+## Alternatives
+
+- **Gradually rewrite current base classes.** We could attempt to gradually rewrite
+Ember constructs such as Services, Routes, and Components in place. This would
+give us more flexibility for a lot of things, like constructor behavior and
+default method implementations, etc. It would probably take about as long to do
+this as the delegate pattern, and could get very messy along the way, especially
+for removing certain constructs (like Mixins).
+- **Provide new base classes at the same time as the old ones.** We could provide
+new base classes, exposed at different import paths, that users could gradually
+switch to over time. This would give us more flexibility as well, but it would
+also be about as much work and the same adoption timeline as the delegate
+pattern, since Ember would have to support both the old Ember object model and
+the new base classes at the same time until the switch. It would also likely be
+fairly confusing to users, since they would have to make sure they were
+importing the correct classes.
+
+## Unresolved questions
+
+If we go this direction, it means we cannot specify any base class behavior. One
+issue that arises is the fact that dependency injections will not be available
+to the class _during_ object construction. They can be assigned _after_ only.
+
+There are a few options for how we could resolve this:
+
+1. **Provide injected properties as named params in the constructor.** This would
+allow us to mirror the injection definitions, which is fairly intuitive.
+However, accessors and methods which depend on services will not work unless
+the user assigns the service, which may be counterintuitive:
+
+  ```js
+  class Foo {
+    @service time;
+
+    getCurrentTime() {
+      return this.time.now();
+    }
+
+    constructor({ time }) {
+      this.currentTime = time.now(); // this works
+      this.currentTime = this.getCurrentTime(); // this does not
+    }
+  }
+  ```
+
+2. **Provide injected properties as positional params to the constructor, using a
+separate decorator.** This would completely separate named injections from
+constructor injections entirely. This would likely be confusing to newcomers,
+but would be similar to how other DI systems work:
+
+  ```js
+  @inject('time', 'date')
+  class Foo {
+    @service time;
+
+    constructor(time, date) {
+      this.currentTime = time.now();
+      this.currentDate = date.today();
+    }
+  }
+  ```
+
+
+3. **Utilize private global state to provide the container to injections during
```

Comment:
I like this, but then it should be done for `getOwner()` so that even a manual call to `getOwner(this)` inside the constructor will work.

Reply:
I agree, this would make the most sense if we pursued this strategy. Will update the RFC to reflect that